pyro/distributions/spanning_tree.py

```diff
@@ -0,0 +1,418 @@
+        raise ValueError('PyTorch cannot handle zero-sized multidimensional tensors')
+    V = num_vertices
+    K = V * (V - 1) // 2
+    v1 = torch.arange(V)
+    v2 = torch.arange(V).unsqueeze(-1)
+    v1, v2 = torch.broadcast_tensors(v1, v2)
+    v1 = v1.contiguous().view(-1)
+    v2 = v2.contiguous().view(-1)
+    mask = (v1 < v2)
+    grid = torch.stack((v1[mask], v2[mask]))
+    assert grid.shape == (2, K)
+    return grid
+
+
+def _remove_edge(grid, e2k, neighbors, components, e):
+    """
+    Remove an edge from a spanning tree.
+    """
+    k = e2k[e]
+    v1 = grid[0, k].item()
+    v2 = grid[1, k].item()
+    neighbors[v1].remove(v2)
+    neighbors[v2].remove(v1)
+    components[v1] = 1
+    pending = [v1]
+    while pending:
+        v1 = pending.pop()
+        for v2 in neighbors[v1]:
+            if not components[v2]:
+                components[v2] = 1
+                pending.append(v2)
+    return k
+
+
+def _add_edge(grid, e2k, neighbors, components, e, k):
+    """
+    Add an edge connecting two components to create a spanning tree.
+    """
+    e2k[e] = k
+    v1 = grid[0, k].item()
+    v2 = grid[1, k].item()
+    neighbors[v1].add(v2)
+    neighbors[v2].add(v1)
+    components.fill_(0)
+
+
+def _find_valid_edges(components, valid_edges):
+    """
+    Find all edges between two components in a complete undirected graph.
+
+    :param components: A [V]-shaped array of boolean component ids. This
+        assumes there are exactly two nonemtpy components.
+    :param valid_edges: An uninitialized array where output is written. On
+        return, the subarray valid_edges[:end] will contain edge ids k for all
+        valid edges.
+    :returns: The number of valid edges found.
+    """
+    k = 0
+    end = 0
+    for v2, c2 in enumerate(components):
+        for v1 in range(v2):
+            if c2 ^ components[v1]:
+                valid_edges[end] = k
+                end += 1
+            k += 1
+    return end
+
+
+@torch.no_grad()
+def _sample_tree_mcmc(edge_logits, edges):
+    if len(edges) <= 1:
+        return edges
+
+    E = len(edges)
+    V = E + 1
+    K = V * (V - 1) // 2
+    grid = make_complete_graph(V)
+    e2k = torch.empty(E, dtype=torch.long)
+    neighbors = {v: set() for v in range(V)}
+    components = torch.zeros(V, dtype=torch.uint8)
+    for e in range(E):
+        v1, v2 = map(int, edges[e])
+        assert v1 < v2
+        e2k[e] = v1 + v2 * (v2 - 1) // 2
+        neighbors[v1].add(v2)
+        neighbors[v2].add(v1)
+    valid_edges_buffer = torch.empty(K, dtype=torch.long)
+
+    for e in range(E):
+        k = _remove_edge(grid, e2k, neighbors, components, e)
+        num_valid_edges = _find_valid_edges(components, valid_edges_buffer)
+        valid_edges = valid_edges_buffer[:num_valid_edges]
+        valid_logits = edge_logits[valid_edges]
+        valid_probs = (valid_logits - valid_logits.max()).exp()
+        total_prob = valid_probs.sum()
+        if total_prob > 0:
```

Comment:
can this not be true?

Reply:
Yes, this can happen when all `valid_logits` are `-inf`. If that's the case the mcmc kernel will avoid moving rather than erroring. I added this a couple years ago to avoid errors in real datasets.